sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -588,9 +596,12 @@ class RocksDB(
           // inside the uploadSnapshot() called below.
           // If changelog checkpointing is enabled, snapshot will be uploaded asynchronously
           // during state store maintenance.
-          latestSnapshot.foreach(_.close())
+          oldSnapshots += latestSnapshot
           latestSnapshot = Some(
-            RocksDBSnapshot(checkpointDir, newVersion, numKeysOnWritingVersion))
+            RocksDBSnapshot(checkpointDir,
+              newVersion,
+              numKeysOnWritingVersion,
+              fileManager.captureFileMapReference()))
```

Comment:
I think we need to synchronize this part to prevent race with maintenance thread on mutating latestSnapshot.